### PR TITLE
REF&MACRO: support rename refactoring of items expanded from a macro

### LINKED ISSUE #5740: Rename inside a macro invocation fails
## Environment

* **IntelliJ Rust plugin version:** 0.3.126.3220-201
* **Rust toolchain version:** 1.46
* **IDE name and version:** CLion 2020.1
* **Operating system:** Windows 10

## Problem description

Renaming an identifier inside a macro invocation fails with `some generated file name.rs is read-only` error.

## Steps to reproduce

```rust
macro_rules! test {
    ($name:ident) => {
        pub static $name: i32 = 0;
    }
}

test!(SYMBOL); // try renaming SYMBOL
```

### LINKED ISSUE #8003: Refactoring types with attributes brittle since last update
## Environment

* **IntelliJ Rust plugin version:** 0.4.158.4203-212
* **Rust toolchain version:** 1.58.0-nightly (00d5e42e7 2021-10-24) x86_64-pc-windows-msvc
* **IDE name and version:** CLion 2021.2.3 (CL-212.5457.51)
* **Operating system:** Windows 10 10.0
* **Macro expansion engine:** new
* **Name resolution engine:** new

I have all "Experimental Features" enabled, esp. for proc macros.

## Problem description

First of all, the new attribute macro expansion is great and it makes debugging attributes much easier. Unfortunately though, refactoring has become rather brittle.

When I try to rename this type, e.g., `MeshDataPostal` to `MeshDataPostalX`, I receive errors like these:

image

These errors _usually_ disappear when I disable `#[ffi_type]` before renaming, but not always. 

FYI, the attribute macro above expanded to:

```
#[repr(C)]
#[derive(Debug, Copy, Default, Clone, Serialize, Deserialize, PartialEq)]
pub struct MeshDataPostal {
    pub mesh_id: MeshID,
    pub num_vertices: u32,
    pub num_indices: u32,
}
unsafe impl interoptopus::lang::rust::CTypeInfo for MeshDataPostal {
    fn type_info() -> interoptopus::lang::c::CType {
        let documentation = interoptopus::lang::c::Documentation::from_line("");
        let mut meta = interoptopus::lang::c::Meta::with_namespace_documentation("".to_string(), documentation);
        let mut fields: std::vec::Vec<interoptopus::lang::c::Field> = std::vec::Vec::new();
        let mut generics: std::vec::Vec<String> = std::vec::Vec::new();
        let name = format!("{}{}", "MeshDataPostal".to_string(), generics.join(""));
        {
            let documentation = interoptopus::lang::c::Documentation::from_line("");
            let the_type = <MeshID as interoptopus::lang::rust::CTypeInfo>::type_info();
            let field = interoptopus::lang::c::Field::with_documentation("mesh_id".to_string(), the_type, interoptopus::lang::c::Visibility::Public, documentation);
            fields.push(field);
        }
        {
            let documentation = interoptopus::lang::c::Documentation::from_line("");
            let the_type = <u32 as interoptopus::lang::rust::CTypeInfo>::type_info();
            let field = interoptopus::lang::c::Field::with_documentation("num_vertices".to_string(), the_type, interoptopus::lang::c::Visibility::Public, documentation);
            fields.push(field);
        }
        {
            let documentation = interoptopus::lang::c::Documentation::from_line("");
            let the_type = <u32 as interoptopus::lang::rust::CTypeInfo>::type_info();
            let field = interoptopus::lang::c::Field::with_documentation("num_indices".to_string(), the_type, interoptopus::lang::c::Visibility::Public, documentation);
            fields.push(field);
        }
        let rval = interoptopus::lang::c::CompositeType::with_meta(name, fields, meta);
        interoptopus::lang::c::CType::Composite(rval)
    }
}
```

### LINKED ISSUE #8182: Proc macro interferes with rename refactor and Go to Declaration or Usage, with experimental features on
## Environment

* **IntelliJ Rust plugin version:** 0.4.160.4261-212
* **Rust toolchain version:** 1.56.1
* **IDE name and version:** CLion CL-212.4746.93 July 27, 2021
* **Operating system:** Windows

I have proc macro and build script experimental features on.

## Problem description

The identifier `App` on line 3 is the same as the one on line 10. IDE recognizes they are the same, as shown by the syntax highlighting. "Go to Declaration or Usage" (by ctrl+click) finds one from the other.

image

But Rename refactor action on `App` fails with an error message:
image

I am positive the renaming refactor worked roughly a week ago. This might be caused by the 11/22 rust plugin update.

Also, find usages and declaration only work partially correct under a proc macro. Detailed tests and comments are below.

Worth noting, macro expansion intention works on all the proc macros.

### Cargo.toml
```toml
[package]
name = "irrelevant"
version = "0.1.0"
edition = "2021"

[dependencies]
yew = { git = "https://github.com/yewstack/yew/" }
gloo-console = "0.2.1"
```

### main.rs

```rust
use yew::prelude::*;

#[function_component(App)]
fn app() -> Html {

    // inside the function_component proc macro,
    // "Go to Declaration or Usage" unilaterally works for declaration, but fails for usage
    // ctrl+click on the first line shows "No usage found in all places"
    // ctrl+click on the following two line redirects cursor back to the first line as expected
    // Renaming `a` on the first line does nothing to the following lines
    // Renaming `a` on the following two lines fails with read-only file error.
    let a = 1;
    gloo_console::console_dbg!(a);
    gloo_console::console_dbg!(a + 1);

    // renaming `a` here this works all as expected.
    // However, ctrl-click on the first `a` says "No usages found in all places"
    let a = 1;
    let b = vec![a];

    html! {
    }
}

fn main() {

    // in a non proc macro context, everything works expectedly
    let a = 1;
    gloo_console::console_dbg!(a);
    gloo_console::console_dbg!(a + 1);

    // in a non proc macro context, everything works expectedly
    let a = 1;
    let b = vec![a];

    yew::start_app::<App>();
}
```

### LINKED ISSUE #8880: Rename doesn't work for macro identifiers
## Environment

* Plugin: 0.4.171.4621-221-nightly
* Rust: rustc 1.63.0-nightly (42e1761c7 2022-05-15)
* IDE: IntelliJ IDEA 2022.1.1 (Community Edition) Build #IC-221.5591.52, built on May 10, 2022
* OS: Macos 12.4

## Problem description

Renaming identifiers created in macro declaration doesn't work

## Steps to reproduce

1. 
```rust
mod foo {
    macro_rules! include_test_data {
        ($name:ident, $val:literal) => {
            pub const $name: u64 = $val;
        };
    }
    include_test_data!(D14094050, 14094050);
}

fn main() {
    use crate::foo::D14094050;

    println!("Hello, world! {}", D14094050);
}
```
2. 
img
3. 
img

### PR DESCRIPTION
**Pull request by vlad20012**
Fixes #5740, fixes #8003, fixes #8880, fixes #8182, probably fixes #8182

changelog: support rename refactoring of items expanded from a macro

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/refactoring/RsRenameProcessor.kt [hunk 1]
```diff
@@ -35,7 +41,7 @@ class RsRenameProcessor : RenamePsiElementProcessor() {
         }
     }
 
-    override fun canProcessElement(element: PsiElement): Boolean = element is RsNamedElement
+    override fun canProcessElement(element: PsiElement): Boolean = element is RsNamedElement || element is RsFakeMacroExpansionRenameablePsiElement
```
Comment (dima74) [inline src/main/kotlin/org/rust/ide/refactoring/RsRenameProcessor.kt:44]:
```suggestion
    override fun canProcessElement(element: PsiElement): Boolean =
        element is RsNamedElement || element is RsFakeMacroExpansionRenameablePsiElement
```

File: src/main/kotlin/org/rust/ide/refactoring/RsRenameProcessor.kt [hunk 2]
```diff
@@ -126,12 +132,42 @@ class RsRenameProcessor : RenamePsiElementProcessor() {
...
+        val superElement = (element as? RsAbstractable)?.superItem ?: element
+        return superElement.findFakeElementForRenameInMacroBody() ?: superElement
+    }
+
+    private fun PsiElement.findFakeElementForRenameInMacroBody(): PsiElement? {
+        if (this is RsNameIdentifierOwner) {
+            val identifier = nameIdentifier
+            val sourceIdentifier = identifier?.findElementExpandedFrom()
+            if (sourceIdentifier != null) {
+                when (val sourceIdentifierParent = sourceIdentifier.parent) {
```
Comment (dima74) [inline src/main/kotlin/org/rust/ide/refactoring/RsRenameProcessor.kt:145]: Is it possible to handle case when `sourceIdentifierParent` is `RsPath`? Like in #8182, when renaming parameter of attr proc macro:

```rust
#[attr(/*caret*/Foo)]
fn func() { ... }
```

**Comment by dima74** [comment]
👍
bors r+

**Comment by vlad20012** [comment]
bors r=dima74

**Pull request merged**